### ISSUE TITLE
Performance regression causes large slowdowns

### ISSUE THREAD

**Issue by matthewfeickert** [diana-hep/pyhf#572]
# Description

A performance regression was introduced in PR #553 that caused the benchmark tests with the Minuit optimizer to take too long for 200 bins and cause the CI systems to timeout and fail, which motivated PR #571.

The regression should be understood and fixed if possible and then the components of PR #571 that addressed this should be reverted.

# Expected Behavior

The speed of evaluation across all backends should remain unchanged from batching and simultaneous evaluation.

# Actual Behavior

The Minuit optimizer for the NumPy backend experiences significant enough slowdowns that it causes the CI systems to timeout.

# Steps to Reproduce

Run the benchmarks in the CI system.

# Checklist

- [x] Run `git fetch` to get the most up to date version of `master`
- [x] Searched through existing Issues to confirm this is not a duplicate issue
- [x] Filled out the Description, Expected Behavior, Actual Behavior, and Steps to Reproduce sections above or have edited/removed them in a way that fully describes the issue

**Comment by lukasheinrich** [comment]
just to set the stage, there are a few commits that are relevant

```
86269a0 feat: Simultaneous pdf class (#553)
82a8bf0 Add probability module (#551)
657cd2f Rewrite Model into PDF composition of MainModel and ConstrainedModel (#549)
9b93f9d Enable full batch evaluation (#546)
673c1b6 Add ParamViewer class  (#519) 
c8a94f4 Add 'tile' method to tensorlib API (#538) << baseline 
```

the number of function calls according to cProfile to evaluate a pdf 100 times

```
import pyhf
import json
import cProfile
w = pyhf.Workspace(json.load(open('./s+b.json')))
m = w.model()
d = w.data(m)
cProfile.run('[m.logpdf(m.config.suggested_init(),d) for x in range(100)] ')
```

```
HEAD is now at 86269a0 feat: Simultaneous pdf class (#553)
         42304 function calls (42104 primitive calls) in 0.050 seconds
Previous HEAD position was 86269a0 feat: Simultaneous pdf class (#553)
HEAD is now at 82a8bf0 Add probability module (#551)
         34504 function calls in 0.035 seconds
Previous HEAD position was 82a8bf0 Add probability module (#551)
HEAD is now at 657cd2f Rewrite Model into PDF composition of MainModel and ConstrainedModel (#549)
         28704 function calls in 0.031 seconds
Previous HEAD position was 657cd2f Rewrite Model into PDF composition of MainModel and ConstrainedModel (#549)
HEAD is now at 9b93f9d Enable full batch evaluation (#546)
         28504 function calls in 0.034 seconds
Previous HEAD position was 9b93f9d Enable full batch evaluation (#546)
HEAD is now at 673c1b6 Add ParamViewer class  (#519)
         31104 function calls in 0.036 seconds
Previous HEAD position was 673c1b6 Add ParamViewer class  (#519)
HEAD is now at c8a94f4 Add 'tile' method to tensorlib API (#538)
         20604 function calls in 0.031 seconds
```

**Comment by lukasheinrich** [comment]
this is the number of core tensorlib calls

```
cat criticalcommits |awk '{print $1}'|while read x; do git checkout $x; python debug.py |grep nump_;bacend|awk '{a+=$1}END{print a}';done
```

```
Previous HEAD position was c8a94f4 Add 'tile' method to tensorlib API (#538)
HEAD is now at 86269a0 feat: Simultaneous pdf class (#553)
9000
Previous HEAD position was 86269a0 feat: Simultaneous pdf class (#553)
HEAD is now at 82a8bf0 Add probability module (#551)
7400
Previous HEAD position was 82a8bf0 Add probability module (#551)
HEAD is now at 657cd2f Rewrite Model into PDF composition of MainModel and ConstrainedModel (#549)
6700
Previous HEAD position was 657cd2f Rewrite Model into PDF composition of MainModel and ConstrainedModel (#549)
HEAD is now at 9b93f9d Enable full batch evaluation (#546)
6700
Previous HEAD position was 9b93f9d Enable full batch evaluation (#546)
HEAD is now at 673c1b6 Add ParamViewer class  (#519)
7200
Previous HEAD position was 673c1b6 Add ParamViewer class  (#519)
HEAD is now at c8a94f4 Add 'tile' method to tensorlib API (#538)
6200
```

**Comment by lukasheinrich** [comment]
closing this as it was fixed in #576
